File: src/main/java/no/novari/flyt/history/mapping/TimeFilterMappingService.java
package no.novari.flyt.history.mapping;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import no.novari.flyt.history.model.instance.ActiveTimePeriod;
import no.novari.flyt.history.model.time.ManualTimeFilter;
import no.novari.flyt.history.model.time.OffsetTimeFilter;
import no.novari.flyt.history.model.time.TimeFilter;
import no.novari.flyt.history.repository.filters.TimeQueryFilter;
import org.springframework.stereotype.Service;

import java.time.Clock;
import java.time.DayOfWeek;
import java.time.OffsetDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.temporal.TemporalAdjusters;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;

@Service
public class TimeFilterMappingService {

    private final Clock clock;
    private final Validator validator;

    public TimeFilterMappingService(Clock clock, Validator validator) {
        this.clock = clock;
        this.validator = validator;
    }

    public TimeQueryFilter toQueryFilter(TimeFilter timeFilter, ZoneId zoneId) {
        Set<ConstraintViolation<TimeFilter>> validate = validator.validate(timeFilter);
        if (!validate.isEmpty()) {
            throw new IllegalArgumentException(validate.toString());
        }
        if (zoneId == null) {
            throw new IllegalArgumentException("Zone id must not be null");
        }

        if (Objects.isNull(timeFilter)) {
            return TimeQueryFilter.EMPTY;
        }
        return Optional.ofNullable(timeFilter.getOffset()).map(this::createQueryFilterFromOffsetTimeFilter)
                .or(() -> Optional.ofNullable(timeFilter.getCurrentPeriod())
                        .map(currentPeriod -> createQueryFilterFromCurrentPeriodTimeFilter(currentPeriod, zoneId)))
                .or(() -> Optional.ofNullable(timeFilter.getManual()).map(this::createQueryFilterFromManualTimeFilter))
                .orElse(TimeQueryFilter.EMPTY);
    }

    private TimeQueryFilter createQueryFilterFromOffsetTimeFilter(OffsetTimeFilter offsetTimeFilter) {
        OffsetDateTime currentTime = OffsetDateTime.now(clock);
        return TimeQueryFilter.builder()
                .latestStatusTimestampMin(
                        currentTime
                                .minusHours(Optional.ofNullable(offsetTimeFilter.getHours()).orElse(0))
                                .minusMinutes(Optional.ofNullable(offsetTimeFilter.getMinutes()).orElse(0))
                )
                .latestStatusTimestampMax(currentTime)
                .build();
    }

    private TimeQueryFilter createQueryFilterFromCurrentPeriodTimeFilter(ActiveTimePeriod activeTimePeriod, ZoneId zoneId) {
        ZonedDateTime currentZonedTime = OffsetDateTime.now(clock).atZoneSameInstant(zoneId);
        return switch (activeTimePeriod) {
            case TODAY -> TimeQueryFilter
                    .builder()
                    .latestStatusTimestampMin(getStartOfDayInUtc(currentZonedTime))
                    .latestStatusTimestampMax(getEndOfDayInUtc(currentZonedTime))
                    .build();
            case THIS_WEEK -> TimeQueryFilter
                    .builder()
                    .latestStatusTimestampMin(getStartOfDayInUtc(
                            currentZonedTime.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
                    ))
                    .latestStatusTimestampMax(getEndOfDayInUtc(
                            currentZonedTime.with(TemporalAdjusters.nextOrSame(DayOfWeek.SUNDAY))
                    ))
                    .build();
            case THIS_MONTH -> TimeQueryFilter
                    .builder()
                    .latestStatusTimestampMin(getStartOfDayInUtc(
                            currentZonedTime.with(TemporalAdjusters.firstDayOfMonth())
                    ))
                    .latestStatusTimestampMax(getEndOfDayInUtc(
                            currentZonedTime.with(TemporalAdjusters.lastDayOfMonth())
                    ))
                    .build();
            case THIS_YEAR -> TimeQueryFilter
                    .builder()
                    .latestStatusTimestampMin(getStartOfDayInUtc(
                            currentZonedTime.with(TemporalAdjusters.firstDayOfYear())
                    ))
                    .latestStatusTimestampMax(getEndOfDayInUtc(
                            currentZonedTime.with(TemporalAdjusters.lastDayOfYear())
                    ))
                    .build();
        };
    }

    private OffsetDateTime getStartOfDayInUtc(ZonedDateTime zonedDateTime) {
        return OffsetDateTime.of(
                zonedDateTime
                        .toLocalDate()
                        .atStartOfDay(),
                zonedDateTime.getOffset()
        ).withOffsetSameInstant(ZoneOffset.UTC);
    }

    private OffsetDateTime getEndOfDayInUtc(ZonedDateTime zonedDateTime) {
        return OffsetDateTime.of(
                zonedDateTime
                        .toLocalDate()
                        .plusDays(1).atStartOfDay(),
                zonedDateTime.getOffset()
        ).withOffsetSameInstant(ZoneOffset.UTC);
    }

    private TimeQueryFilter createQueryFilterFromManualTimeFilter(ManualTimeFilter manualTimeFilter) {
        return TimeQueryFilter
                .builder()
                .latestStatusTimestampMin(
                        Optional.ofNullable(manualTimeFilter.getMin())
                                .map(t -> t.withOffsetSameInstant(ZoneOffset.UTC))
                                .orElse(null)
                )
                .latestStatusTimestampMax(Optional.ofNullable(manualTimeFilter.getMax())
                        .map(t -> t.withOffsetSameInstant(ZoneOffset.UTC))
                        .orElse(null)
                )
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/history/mapping/TimeFilterMappingService.kt
package no.novari.flyt.history.mapping

import jakarta.validation.Validator
import no.novari.flyt.history.model.instance.ActiveTimePeriod
import no.novari.flyt.history.model.time.ManualTimeFilter
import no.novari.flyt.history.model.time.OffsetTimeFilter
import no.novari.flyt.history.model.time.TimeFilter
import no.novari.flyt.history.repository.filters.TimeQueryFilter
import org.springframework.stereotype.Service
import java.time.Clock
import java.time.DayOfWeek
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.temporal.TemporalAdjusters

@Service
class TimeFilterMappingService(
    private val clock: Clock,
    private val validator: Validator,
) {
    fun toQueryFilter(
        timeFilter: TimeFilter?,
        zoneId: ZoneId,
    ): TimeQueryFilter {
        if (timeFilter == null) {
            return TimeQueryFilter.EMPTY
        }

        val violations = validator.validate(timeFilter)
        require(violations.isEmpty()) { violations.toString() }

        return when {
            timeFilter.offset != null -> {
                createQueryFilterFromOffsetTimeFilter(timeFilter.offset)
            }

            timeFilter.currentPeriod != null -> {
                createQueryFilterFromCurrentPeriodTimeFilter(
                    timeFilter.currentPeriod,
                    zoneId,
                )
            }

            timeFilter.manual != null -> {
                createQueryFilterFromManualTimeFilter(timeFilter.manual)
            }

            else -> {
                TimeQueryFilter.EMPTY
            }
        }
    }

    private fun createQueryFilterFromOffsetTimeFilter(offsetTimeFilter: OffsetTimeFilter): TimeQueryFilter {
        val currentTime = OffsetDateTime.now(clock)
        return TimeQueryFilter
            .builder()
            .latestStatusTimestampMin(
                currentTime
                    .minusHours(offsetTimeFilter.hours?.toLong() ?: 0)
                    .minusMinutes(offsetTimeFilter.minutes?.toLong() ?: 0),
            ).latestStatusTimestampMax(currentTime)
            .build()
    }

    private fun createQueryFilterFromCurrentPeriodTimeFilter(
        activeTimePeriod: ActiveTimePeriod,
        zoneId: ZoneId,
    ): TimeQueryFilter {
        val currentZonedTime = OffsetDateTime.now(clock).atZoneSameInstant(zoneId)

        return when (activeTimePeriod) {
            ActiveTimePeriod.TODAY -> {
                TimeQueryFilter
                    .builder()
                    .latestStatusTimestampMin(getStartOfDayInUtc(currentZonedTime))
                    .latestStatusTimestampMax(getEndOfDayInUtc(currentZonedTime))
                    .build()
            }

            ActiveTimePeriod.THIS_WEEK -> {
                TimeQueryFilter
                    .builder()
                    .latestStatusTimestampMin(
                        getStartOfDayInUtc(currentZonedTime.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))),
                    ).latestStatusTimestampMax(
                        getEndOfDayInUtc(currentZonedTime.with(TemporalAdjusters.nextOrSame(DayOfWeek.SUNDAY))),
                    ).build()
            }

            ActiveTimePeriod.THIS_MONTH -> {
                TimeQueryFilter
                    .builder()
                    .latestStatusTimestampMin(
                        getStartOfDayInUtc(currentZonedTime.with(TemporalAdjusters.firstDayOfMonth())),
                    ).latestStatusTimestampMax(
                        getEndOfDayInUtc(currentZonedTime.with(TemporalAdjusters.lastDayOfMonth())),
                    ).build()
            }

            ActiveTimePeriod.THIS_YEAR -> {
                TimeQueryFilter
                    .builder()
                    .latestStatusTimestampMin(
                        getStartOfDayInUtc(currentZonedTime.with(TemporalAdjusters.firstDayOfYear())),
                    ).latestStatusTimestampMax(
                        getEndOfDayInUtc(currentZonedTime.with(TemporalAdjusters.lastDayOfYear())),
                    ).build()
            }
        }
    }

    private fun getStartOfDayInUtc(zonedDateTime: ZonedDateTime): OffsetDateTime {
        return OffsetDateTime
            .of(zonedDateTime.toLocalDate().atStartOfDay(), zonedDateTime.offset)
            .withOffsetSameInstant(ZoneOffset.UTC)
    }

    private fun getEndOfDayInUtc(zonedDateTime: ZonedDateTime): OffsetDateTime {
        return OffsetDateTime
            .of(zonedDateTime.toLocalDate().plusDays(1).atStartOfDay(), zonedDateTime.offset)
            .withOffsetSameInstant(ZoneOffset.UTC)
    }

    private fun createQueryFilterFromManualTimeFilter(manualTimeFilter: ManualTimeFilter): TimeQueryFilter {
        return TimeQueryFilter
            .builder()
            .latestStatusTimestampMin(manualTimeFilter.min?.withOffsetSameInstant(ZoneOffset.UTC))
            .latestStatusTimestampMax(manualTimeFilter.max?.withOffsetSameInstant(ZoneOffset.UTC))
            .build()
    }
}
